src/fm/utils.cljc

```diff
@@ -328,21 +333,28 @@
   (if (empty? (meta args-form))
     (fn-form form-args)
 
-    (let [metadata (into {} (map meta-xf) (meta args-form))
-          bindings (interleave
-                    (map :fm.meta/sym  (vals metadata))
-                    (map :fm.meta/form (vals metadata)))
-          fn-form  (fn-form
-                    (merge
-                     form-args
-                     {:fm/metadata metadata}))
-          meta     (not-empty
-                    (zipmap
-                     (keys metadata)
-                     (map :fm.meta/sym (vals metadata))))]
-
-      `(let [~@bindings]
-         (with-meta ~fn-form ~meta)))))
+    (let [metadata              (into {} (map meta-xf) (meta args-form))
+          metadata-vals         (vals metadata)
+          metadata-key-bindings (map :fm.meta/sym metadata-vals)
+          bindings              (interleave
+                                 metadata-key-bindings
+                                 (map :fm.meta/form metadata-vals))
+          fn-form               (fn-form
+                                 (merge
+                                  form-args
+                                  {:fm/metadata metadata}))
+          symbol-meta           (when (:fm/doc metadata)
+                                  {:doc (get-in
+                                         metadata
+                                         [:fm/doc :fm.meta/form])})
+          meta                  (not-empty
+                                 (zipmap
+                                  (keys metadata)
+                                  metadata-key-bindings))]
+
+      [`(let [~@bindings]
+          (with-meta ~fn-form ~meta))
+       symbol-meta])))
```

Comment:
Definitely open to something different here, feel kinda gross having to pass this back, but couldn't figure out a better way to extract `:fm/doc` than to pre-compute it here and pass it to the macro.

Reply:
Passing more back from `fm-form` isn't gross 😂 but I think I'd have passed back a map with `::form` and `::meta` or similar instead of going positional. I'd also probably change the fn name to `fm-context` or `fm-data` or something like that.